Sources/Fuzzilli/FuzzIL/Analyzer.swift

```diff
@@ -142,11 +142,16 @@ struct VariableAnalyzer: Analyzer {
 /// Keeps track of the current context during program construction.
 struct ContextAnalyzer: Analyzer {
     private var contextStack = Stack([Context.javascript])
+    private var breakContextStack = Stack([Context.empty])
 
     var context: Context {
         return contextStack.top
     }
 
+    var breakContext: Context {
```

Comment:
Is it enough if you check if `.context.contains(.loop)` (and otherwise if it contains .switch, and otherwise print an error)? Then I think you don't need any changes to this Analyzer and to Instruction.swift at all.

Reply:
Ok no that doesn't work since you can have both context active. So I think you'd need to iterate over `contextStack` until you find either a .loop or a .switch context, and then whichever you find first is the one you care about? I think it's fine to make the contextStack `private(set) var contextStack` for that (i.e. readable from outside, but not writable)

Reply:
That was my first approach also but I found a problem with that:

If we are in a context which contains both .loop and .switchBlock we need to determine which one was most recently opened. We can try to determine this by iterating over the stack (most recent first) and determine the first context where either .loop or .switchBlock disappears. For example in this case it works
```
loop
    switch
        break 
```
switch was opened more recently so it will disappear first from the context stack. In this case it works. And honestly in 99% of cases it is fine. But we can also have something like this:

```
loop
    switch
        loop
            break
```
The first context that starts to disappear is from the switch because the loop context opened from the outer loop is still open even when we are leaving the inner loop.

Reply:
I think it works if you instead look for the first context where  either .loop or .switchBlock _appears_. So in
```
loop
    switch
        loop
            break
```
You would iterate backwards (starting at the innermost context). And then you'd find .loop before .switch, and so you know the break must be a LoopBreak. Does that work?

Reply:
That's exactly the idea behind the newly added breakContext stack. The problem with the contextStack in contextAnalyzer is that the contexts are option sets and so the innermost context is not just the loop but the union of all all contexts that envelop it. 
For example in the example above context would simply be "2113" so in other words 2^11 xor 2^6 xor 2^0, i.e. both a javascript, loop and switchBlock context. Just with this option set it is unclear which one has precedence. 
The breakContextStack was added to have this precedence information

Reply:
Ah right, makes sense. I wonder if we should just do something like this when seeing a `break` in the compiler:
```
// Determine which type of break we have.
// Note that this is currently slightly incorrect for code such as
//
//     while {
//         switch {
//             case X {
//                  break;
//
// However, this is probably actually a bug in our context analysis: the BeginSwitchCase should not continue the outer .loop 
// context as a LoopBreak should not be allowed inside of it. Once that is fixed, this code will be fully correct (for now,
// it is already "correct enough" in the sense that compiling and lifting results in a semantically equivalent program).
if contextAnalyzer.context.contains(.loop) {
    emit(LoopBreak())
} else if contextAnalyzer.context.contains(.switch) {
    emit(SwitchBreak())
} else {
    error();
}
```

I'd like to keep these changes "local" to the compiler and not affect other parts of Fuzzilli. The compiler runs ahead of time, so isn't really performance sensitive, so it'd be nice if it does't require changes to the Fuzzilli base that cause overhead during fuzzing. IIUC, this should be "good enough" in the sense that compiling and lifting will yield an equivalent program (because both types of breaks will just lift to a `break` statement). Is that correct? As I describe in the comment, I think the current treatment of switch-inside-a-loop is actually buggy. If we fix that, this code will be fully correct.

Reply:
Good idea. In the most recent commit I've made a change where the bug you mentioned is fixed. This also moves all the logic for break context identification out of the performance critical parts of Fuzzilli and into the ContextAnalyzer. I think the way it is implemented at the moment reflects closely what is described in your comment :)